### PR TITLE
[Feature]: Support mmpose

### PR DESCRIPTION
**Pull request by RunningLeon**
## Motivation

Support mmpose deployment with following models: HRNet, LiteHRNet, MSPN

- [configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w48_coco_256x192.py](https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w48_coco_256x192.py)
- [configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/litehrnet_30_coco_256x192.py](https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/litehrnet_30_coco_256x192.py)
- [configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/4xmspn50_coco_256x192.py](https://github.com/open-mmlab/mmpose/blob/master/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/4xmspn50_coco_256x192.py)
## Modification

- [x] add mmpose.deploy
- [x] add rewritings for mmpose models
- [x] add mmpose deploy configs
- [x] add unittests
- [x] add docs & update docs
- [x] align with pytorch

<div style="margin-left: 25px;">
<table class="docutils">
<thead>
  <tr>
    <th align="center" colspan="4">MMpose</th>
    <th align="center">Pytorch</th>
    <th align="center">ONNXRuntime</th>
    <th align="center" colspan="3">TensorRT</th>
    <th align="center">PPLNN</th>
    <th align="center">OpenVINO</th>
    <th align="center">Model Config</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td align="center">Model</td>
    <td align="center">Task</td>
    <td align="center">Dataset</td>
    <td align="center">Metrics</td>
    <td align="center">fp32</td>
    <td align="center">fp32</td>
    <td align="center">fp32</td>
    <td align="center">fp16</td>
    <td align="center">int8</td>
    <td align="center">fp16</td>
    <td align="center">fp32</td>
    <td>model config file</td>
  </tr>
  <tr>
    <td align="center" rowspan="2">HRNet</td>
    <td align="center" rowspan="2">Pose Detection</td>
    <td align="center" rowspan="2">COCO</td>
    <td align="center">AP</td>
    <td align="center">0.748</td>
    <td align="center">0.748</td>
    <td align="center">0.748</td>
    <td align="center">0.748</td>
    <td align="center">-</td>
    <td align="center">-</td>
    <td align="center">0.748</td>
    <td rowspan="2">$MMPOSE_DIR/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w48_coco_256x192.py</td>
  </tr>
  <tr>
    <td align="center">AR</td>
    <td align="center">0.802</td>
    <td align="center">0.802</td>
    <td align="center">0.802</td>
    <td align="center">0.802</td>
    <td align="center">-</td>
    <td align="center">-</td>
    <td align="center">0.802</td>
  </tr>
  <tr>
    <td align="center" rowspan="2">LiteHRNet</td>
    <td align="center" rowspan="2">Pose Detection</td>
    <td align="center" rowspan="2">COCO</td>
    <td align="center">AP</td>
    <td align="center">0.663</td>
    <td align="center">0.663</td>
    <td align="center">0.663</td>
    <td align="center">-</td>
    <td align="center">-</td>
    <td align="center">-</td>
    <td align="center">0.663</td>
    <td rowspan="2">$MMPOSE_DIR/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/litehrnet_30_coco_256x192.py</td>
  </tr>
  <tr>
    <td align="center">AR</td>
    <td align="center">0.728</td>
    <td align="center">0.728</td>
    <td align="center">0.728</td>
    <td align="center">-</td>
    <td align="center">-</td>
    <td align="center">-</td>
    <td align="center">0.728</td>
  </tr>
  <tr>
    <td align="center" rowspan="2">MSPN </td>
    <td align="center" rowspan="2">Pose Detection</td>
    <td align="center" rowspan="2">COCO</td>
    <td align="center">AP</td>
    <td align="center">0.762</td>
    <td align="center">0.762</td>
    <td align="center">0.762</td>
    <td align="center">0.762</td>
    <td align="center">-</td>
    <td align="center">-</td>
    <td align="center">0.762</td>
    <td rowspan="2">$MMPOSE_DIR/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/4xmspn50_coco_256x192.py</td>
  </tr>
  <tr>
    <td align="center">AR</td>
    <td align="center">0.825</td>
    <td align="center">0.825</td>
    <td align="center">0.825</td>
    <td align="center">0.825</td>
    <td align="center">-</td>
    <td align="center">-</td>
    <td align="center">0.825</td>
  </tr>
</tbody>
</table>
</div>

### Note
- `flip_test` is explicitly set to `False` for evaluation.

## BC-breaking (Optional)

None

## Use cases (Optional)
1. clone and install [mmpose](https://github.com/open-mmlab/mmpose).
2. conversion to tensorrt for hrnet `configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w48_coco_256x192.py`
```
python tools/deploy.py \
configs/mmpose/posedetection_tensorrt_static-256x192.py \
../mmpose/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w48_coco_256x192.py \
../mmpose/checkpoints/hrnet_w48_coco_256x192-b9e0b3ab_20200708.pth \
demo/resources/human-pose.jpg \
--work-dir work-dirs/mmpose/topdown/hrnet/trt \
--device cuda
```

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### note
mmpose and mmedit conflicts at certain case because the register same name class `DistributedDataParallelWrapper`
This bug can be reproduced with this line
```
python -c 'from mmedit.apis import single_gpu_test; import mmpose.apis.inference'
```

### CODE REVIEW

File: docs/en/codebases/mmpose.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# MMPose Support
+
+[MMPose](https://github.com/open-mmlab/mmpose) is an open-source toolbox for pose estimation based on PyTorch. It is a part of the [OpenMMLab](https://openmmlab.com/) project.
+
+## MMPose installation tutorial
+
+Please refer to [official installation guide](https://github.com/open-mmlab/mmpose/blob/master/docs/en/install.md) to install the codebase.
```
Comment (RunningLeon) [inline docs/en/codebases/mmpose.md:7]: may use readthedocs html link.

File: docs/en/codebases/mmpose.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+## Note
+
+Usually, the mmpose model needs some extra information for input image, but we can't get it directly. So, when exporting the model, you can use `$MMDEPLOY_DIR/demo/resources/human-pose.jpg` as input.
+
+### Example
+
+```bash
+python tools/deploy.py \
+configs/mmpose/posedetection_tensorrt_static-256x192.py \
+../mmpose/configs/body/2d_kpt_sview_rgb_img/topdown_heatmap/coco/hrnet_w48_coco_256x192.py \
```
Comment (RunningLeon) [inline docs/en/codebases/mmpose.md:26]: better to change `../mmpose` to `$MMPOSE_DIR`

File: docs/en/codebases/mmpose.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+../mmpose/checkpoints/hrnet_w48_coco_256x192-b9e0b3ab_20200708.pth \
```
Comment (RunningLeon) [inline docs/en/codebases/mmpose.md:27]: better to change `../mmpose` to `$MMPOSE_DIR`

File: tests/test_codebase/test_mmpose/test_mmpose_models.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    model = TopdownHeatmapMSMUHead((32, 48), unit_channels=2,
+                                   loss_keypoint=dict(type='JointsMSELoss',
+                                                      use_target_weight=False))
+
+    model.requires_grad_(False)
+    return model
+
+
+@pytest.mark.parametrize(
+    'backend_type', [Backend.DEFAULT])
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmpose/test_mmpose_models.py]: default backend will not run inference, better explicitly set backend as onnxruntime and TensorRT 
https://github.com/open-mmlab/mmdeploy/blob/003cf1bfd7cec20f9c296825dab0c2681c3fc5e6/mmdeploy/utils/test.py#L474

File: tests/test_codebase/test_mmpose/test_mmpose_models.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    model = TopdownHeatmapSimpleHead(2, 4, num_deconv_filters=(16, 16, 16),
+                                     loss_keypoint=dict(type='JointsMSELoss',
+                                                        use_target_weight=False
+                                                        ))
+    model.requires_grad_(False)
+    return model
+
+
+@pytest.mark.parametrize(
+    'backend_type', [Backend.DEFAULT])
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmpose/test_mmpose_models.py]: better explicitly set backend

File: tests/test_codebase/test_mmpose/test_mmpose_models.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+    for model_output, rewrite_output in zip(model_outputs, rewrite_outputs):
+        model_output = model_output.cpu().numpy()
+        if isinstance(rewrite_output, torch.Tensor):
+            rewrite_output = rewrite_output.cpu().numpy()
+        assert np.allclose(
+            model_output, rewrite_output, rtol=1e-03, atol=1e-05)
+
+
+@pytest.mark.parametrize(
+    'backend_type', [Backend.DEFAULT])
```
Comment (RunningLeon) [inline tests/test_codebase/test_mmpose/test_mmpose_models.py]: better explicitly set backend

File: mmdeploy/codebase/mmpose/deploy/pose_detection.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+
+        # build the data pipeline
+        channel_order = cfg.test_pipeline[0].get('channel_order', 'rgb')
+        test_pipeline = [LoadImage(channel_order=channel_order)
+                         ] + cfg.test_pipeline[1:]
+        test_pipeline = Compose(test_pipeline)
+        dataset_name = dataset_info.dataset_name
+        flip_pairs = dataset_info.flip_pairs
+        batch_data = []
+        if self.deploy_cfg.onnx_config['input_shape'] is not None:
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmpose/deploy/pose_detection.py:109]: input_shape is passed...

File: mmdeploy/utils/constants.py [hunk 8]
```diff
@@ -24,6 +24,7 @@ class Task(AdvancedEnum):
     CLASSIFICATION = 'Classification'
     OBJECT_DETECTION = 'ObjectDetection'
     INSTANCE_SEGMENTATION = 'InstanceSegmentation'
+    POSE_DETECTION = 'PoseDetection'
```
Comment (grimoire) [inline mmdeploy/utils/constants.py:27]: Should we define different task for 2d/3d kpt?

File: mmdeploy/codebase/mmpose/models/heads/topdown_heatmap_multi_stage_head.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    flip_pairs (None | list[tuple]):
+        Pairs of keypoints which are mirrored.
+
+    Returns:
+        output_heatmap (torch.Tensor): Output heatmaps.
+    """
+    assert flip_pairs is None
+    output = self.forward(x)
+    assert isinstance(output, list)
+    output = output[-1]
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/models/heads/topdown_heatmap_multi_stage_head.py:25]: what is the first dimension of output?

File: mmdeploy/codebase/mmpose/models/backbones/litehrnet.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+def cross_resolution_weighting__forward(ctx, self, x):
+    """Rewrite ``forward`` for default backend.
+
+    Rewrite this function to support export ``adaptive_avg_pool2d``.
+
+    Args:
+        x (list): block input.
+    """
+
+    mini_size = [int(_) for _ in x[-1].shape[-2:]]
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/models/backbones/litehrnet.py:19]: Did that means liteHRNet only support static shape?

File: mmdeploy/codebase/mmpose/deploy/pose_detection_model.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        deploy_cfg (str | mmcv.Config): Model config file or loaded Config
+            object.
+    """
+
+    def __init__(self,
+                 backend: Backend,
+                 backend_files: Sequence[str],
+                 device: str,
+                 deploy_cfg: Union[str, mmcv.Config] = None,
+                 model_cfg: Union[str, mmcv.Config] = None):
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/deploy/pose_detection_model.py:31]: add kwargs here.
You have feed kwargs int `build_pose_detection_model`.

File: mmdeploy/codebase/mmpose/deploy/pose_detection_model.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+        Returns:
+            list: A list contains predictions.
+        """
+        from mmpose.models.heads.topdown_heatmap_base_head import \
+            TopdownHeatmapBaseHead
+
+        input_img = img.contiguous()
+        outputs = self.forward_test(input_img, img_metas, *args, **kwargs)
+        heatmaps = outputs[0]
+        head = TopdownHeatmapBaseHead()
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/deploy/pose_detection_model.py:74]: Why not create it in __init__?

File: mmdeploy/utils/constants.py (hunk 8 above)
Comment (RunningLeon) [reply mmdeploy/utils/constants.py:27]: For now, we did not fully compare the difference of 2d and 3d kpt models. Don't know if it's necessary.

File: mmdeploy/codebase/mmpose/models/backbones/litehrnet.py (hunk 10 above)
Comment (RunningLeon) [reply mmdeploy/codebase/mmpose/models/backbones/litehrnet.py:19]: Yes. We assume all MMPose models do not need to support dynamic shape.

File: mmdeploy/codebase/mmpose/models/heads/topdown_heatmap_simple_head.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+
+from mmdeploy.core import FUNCTION_REWRITER
+
+
+@FUNCTION_REWRITER.register_rewriter(
+    'mmpose.models.heads.TopdownHeatmapSimpleHead.inference_model')
+def top_down_heatmap_simple_head__inference_model(ctx,
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/models/heads/topdown_heatmap_simple_head.py:8]: top_down->topdown

File: mmdeploy/codebase/mmpose/models/heads/topdown_heatmap_simple_head.py [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+                                                  self,
+                                                  x,
+                                                  flip_pairs=None):
+    """Rewrite `forward_test` of TopDown for default backend.
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/models/heads/topdown_heatmap_simple_head.py:12]: You use double semicolon in `topdown_heatmap_multi_stage_head`.

File: mmdeploy/codebase/mmpose/models/backbones/litehrnet.py [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+    out = [F.adaptive_avg_pool2d(s, mini_size) for s in x[:-1]] + [x[-1]]
```
Comment (grimoire) [inline mmdeploy/codebase/mmpose/models/backbones/litehrnet.py:20]: Would it be better to do the concate before adaptive_avg_pool2d?
We only need 1 pool2d node.

Comment (RunningLeon) [reply mmdeploy/codebase/mmpose/models/backbones/litehrnet.py:20]: x is a list of different w, h tensors and cannot be concated before pooling.

File: mmdeploy/codebase/mmpose/deploy/pose_detection.py [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+        from mmpose.datasets.pipelines import Compose
+
+        cfg = self.model_cfg
+
+        dataset_info = cfg.data.test.dataset_info
+        dataset_info = DatasetInfo(dataset_info)
+
+        # create dummy person results
+        if isinstance(imgs, str):
+            height, width = mmcv.imread(imgs).shape[:2]
```
[the remark this answers is not in the document]
Comment (RunningLeon) [reply mmdeploy/codebase/mmpose/deploy/pose_detection.py:95]: agree.

File: configs/mmpose/pose-detection_static.py [hunk 17]
```diff
@@ -0,0 +1,7 @@
+_base_ = ['../_base_/onnx_config.py']
+
+codebase_config = dict(
+    type='mmpose',
+    task='PoseDetection',
+    test_json='',
```
[the remark this answers is not in the document]
Comment (RunningLeon) [reply configs/mmpose/pose-detection_static.py:6]: Yes ,could be removed.

**Comment by RunningLeon** [comment]
Any other comments on this PR?

**Review by grimoire — APPROVED** [review]

**Review by AllentDan — APPROVED** [review]
LGTM

**Pull request merged**